### ISSUE TITLE
Array equal code

### ISSUE THREAD

**Issue by Temtaime** [ldc-developers/ldc#1632]
Currently a == b generates ugly amount of code calling _adEq2 from druntime, involving typeinfo classes, vtbl lookups and cetera...

For example current foo, and optimized foo2 function.
http://goo.gl/0yWheD

I think we should define compareMemory in druntime and call it on POD arrays. And for most cases it will be inlined.

**Comment by JohanEngelen** [comment]
Code with currently terrible codegen:
```d
bool foo(ref uint[2] a, ref uint[2] b)
{
  return a == b;
}
```

**Comment by JohanEngelen** [comment]
Related issue https://github.com/ldc-developers/ldc/issues/1605

**Comment by jairov4** [comment]
Suggestion for better assembly on odd counts
http://goo.gl/bh50y6

**Comment by kinke** [comment]
@jairov4: We need to operate on a higher level here. Firstly, we need to check whether this direct memory comparison is valid - no custom opCmp for the element type, and we need to check for well-defined padding bytes too. It's also questionable whether you want that optimization if the elements contain floating-point values with multiple exact memory representations of a value and their NaN special case.
Only if the optimization is valid, we may lower it to a memcmp() call which should be optimized (also using SSE/AVX registers, not just up to size_t's). I haven't looked at `_adEq2()` so I don't know whether it does have such a fast path already (if so, then most likely at the cost of determining the validness of this optimization at runtime).

**Comment by kinke** [comment]
If one of you guys want to contribute, `gen/arrays.cpp` and function `DtoArrayEquals()` is the place to start from. ;)

**Comment by JohanEngelen** [comment]
`_adEq2` does an indirect call to TypeInfo's comparison function. (which first checks array lengths and then does a memcmp for integer arrays).
Like kinke says, I think the first thing we can easily do here is optimize comparisons of arrays of integer types, by directly calling memcmp for those (we should not reimplement memcmp!). LLVM knows memcpy semantics and replaces the call for _very_ small arrays like `uint[2]`.
Then later, additional logic can be added to check whether this optimization is valid for other types too.

(have a look at what clang does for `std::array<int>` and `std::array<float>` comparisons; for floating point it does not do a memcmp I assume because of what kinke said)

**Comment by kinke** [comment]
> LLVM knows memcpy semantics

But quite oddly, there's no mem**cmp** intrinsic. So we're probably best off by a 'manual' optimization for small arrays (e.g., if castable to a single or a few integers) and otherwise calling the C runtime function directly.

**Comment by JohanEngelen** [comment]
> LLVM knows memcpy semantics

Oops, meant mem**cmp**. So LLVM automatically does the optimization of removing the memcmp call.
See also: http://lists.llvm.org/pipermail/llvm-dev/2010-August/034057.html

**Comment by kinke** [comment]
That proposal sadly never made it into LLVM. But it mentions D. :)

**Comment by JohanEngelen** [comment]
This is quite a fun thing to work on, I think.
Should the call to C-runtime `memcmp` be done at `-O0`, or at some higher optimization level?

**Comment by kinke** [comment]
In my opinion it should always be called, no need to artificially slow down debug builds. :)

**Comment by JohanEngelen** [comment]
Steps suggestion:
- [ ] We have to call the external C-runtime function `memcmp`, so it must be declared. Have a look at how runtime functions are declared in `gen/runtime.cpp`, `buildRuntimeModule()`. (Check the druntime source and make sure that your `memcmp` declaration is correct for all platforms!)
- [ ] Write a function that determines whether the call to `memcmp` is valid for the comparison, i.e. check the types of the left- and right-hand-sides: `bool canCallMemcmp(DValue *l, DValue *r)`, but choose a better name :)   It's perfectly fine if you stay cautious and only allow `int[2]` at first. Allowing other types is easily done later by merely adjusting this function.
- [ ] Add the optimization to `DtoArrayEquals` in `gen/arrays.cpp`. There is already an optimization path for when the RHS is `null`, so you can mimic that.
- [ ] Inspect the outputted LLVM IR (`-c -output-ll`) and add an IR test. It should test whether the memcmp call is inserted when valid, but also test that it is _not_ inserted when it is not valid (choose a non-obvious type, e.g. a `float[]` or a struct[] with padding). Look at `tests/codegen/array_equals_null.d` for a test example, and refer to https://wiki.dlang.org/LDC_Lit-based_testsuite

**Comment by JohanEngelen** [comment]
OK, this was too much fun. I implemented it for static arrays of integers myself.
For static arrays of structs, in the future we can add code for it (checking the validity of using memcmp directly).

**Comment by JohanEngelen** [comment]
#1719

**Comment by mleise** [comment]
I was about to report a performance issue and found this report. I was looking into why my hash table was slower than anything else when looking up `uint[256]` keys. After checking everything else only the comparisons remained and for what it's worth it was *not* `_adEq2`. It was the type info passed into it!

Comparing two `uint[256]` with `_adEq2` is as fast as `memcmp()`, but with `immutable(uint[256])` it is dog slow. I could not reproduce this in a simple test case yet, but the generated assembly for my hash table only differs in the TypeInfo pointer.

Here is a little table to illustrate the problem. The *base line* is where only the hashes are compared and the equality function returns `true`. Since the hashes do not have any collisions this is still valid.

Type used in comparison|Total time looking up 100,000 elements
---|---
base line (always true)|69 ms
memcmp|100 ms (+31 ms)
uint[256]|100 ms (+31 ms)
immutable(uint[256])|217 ms (+148 ms)

In other words: There is more than 100% overhead to the operation of the hash table and a 4.7x slowdown for the comparison operation alone when using immutable types. I can't really make sense of it since a `TypeInfo_Invariant.equals()` just calls its base `TypeInfo.equals()`. If this is a case of a major CPU stall due to nested virtual function calls it would have wider consequences for D, but so far the same code compiled with DMD 2.070 does not show the problem. With both mutable and immutable comparisons the times hover around 200 ms there.
